test_conformance/api/test_queue_array_properties.cpp

```diff
@@ -0,0 +1,273 @@
+﻿//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+
+#include "testBase.h"
+#include "harness/typeWrappers.h"
+#include <sstream>
+#include <string>
+#include <vector>
+#include <algorithm>
+
+struct test_queue_array_properties_data
+{
+    std::vector<cl_queue_properties> properties;
+    std::string description;
+};
+
+int verify_if_properties_supported(
+    cl_device_id deviceID, cl_command_queue_properties requested_bitfield,
+    cl_uint requested_size)
+{
+    int error = CL_SUCCESS;
+    cl_command_queue_properties on_host_properties = 0;
+    cl_device_device_enqueue_capabilities device_enqueue_caps;
+    std::vector<cl_command_queue_properties> all_properties = {
+        CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, CL_QUEUE_PROFILING_ENABLE,
+        CL_QUEUE_ON_DEVICE, CL_QUEUE_ON_DEVICE_DEFAULT
+    };
+
+    error =
+        clGetDeviceInfo(deviceID, CL_DEVICE_QUEUE_ON_HOST_PROPERTIES,
+                        sizeof(on_host_properties), &on_host_properties, NULL);
+    test_error(error,
+               "clGetDeviceInfo for CL_DEVICE_QUEUE_ON_HOST_PROPERTIES failed");
+
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_DEVICE_ENQUEUE_CAPABILITIES,
+                            sizeof(device_enqueue_caps), &device_enqueue_caps,
+                            NULL);
+    test_error(
+        error,
+        "clGetDeviceInfo for CL_DEVICE_DEVICE_ENQUEUE_CAPABILITIES failed");
+    if (!(device_enqueue_caps & CL_DEVICE_QUEUE_SUPPORTED)
+        && requested_size > 0)
+    {
+        log_info("\tCL_DEVICE_QUEUE_SUPPORTED false, skipped\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+    for (auto check_property : all_properties)
+    {
+        if (check_property & requested_bitfield)
+        {
+            if (check_property & on_host_properties)
+            {
+                log_info("\t%s supported\n",
+                         GetQueuePropertyName(check_property));
+            }
+            else if (device_enqueue_caps & CL_DEVICE_QUEUE_SUPPORTED)
+            {
+                if ((check_property & CL_QUEUE_ON_DEVICE)
+                    || (check_property & CL_QUEUE_ON_DEVICE_DEFAULT))
+                {
+                    log_info("\t%s supported\n",
+                             GetQueuePropertyName(check_property));
+                }
+            }
+            else
+            {
+                log_info("\t%s not supported, skipped\n",
```

Comment:
Can we simplify this logic slightly, by:

1. Getting the `CL_DEVICE_QUEUE_ON_DEVICE_PROPERTIES` in addition to the host queue properties.
2. Checking the `requested_bitfield` against the host queue properties or the device queue properties depending whether this is a host queue or a device queue?

    a. I guess we might need to strip off `CL_QUEUE_ON_DEVICE` and `CL_QUEUE_ON_DEVICE_DEFAULT` if we do this since they don't get included in the query for `CL_DEVICE_QUEUE_ON_DEVICE_PROPERTIES`, but that still seems fairly straightforward.

If we do this then I think we can remove the `all_properties` vector and this entire loop.

I think I convinced myself this test logic is OK for now, but it's a bit hard to follow, and without checking the device queue properties I think it could easily become incorrect in the future.

Reply:
I hope this comment is solved in commit https://github.com/KhronosGroup/OpenCL-CTS/pull/869/commits/164f547113d1da99e1672927cf96f44300e11ae2

Reply:
Yup, looks good, thanks!